esp-hal/src/dma/mod.rs

```diff
@@ -1494,24 +1494,261 @@ where
     }
 }
 
-/// Trait to be implemented for an in progress dma transfer.
-#[allow(drop_bounds)]
-#[doc(hidden)]
-pub trait DmaTransfer: Drop {
+pub(crate) mod dma_private {
+    use super::*;
+
+    pub trait DmaSupport {
+        /// Wait until the transfer is done.
+        /// Depending on the peripheral this might include checking the DMA
+        /// channel.
+        fn peripheral_wait_dma(&mut self, is_tx: bool, is_rx: bool);
+
+        /// Only used by circular DMA transfers
+        fn peripheral_dma_stop(&mut self);
+    }
+
+    pub trait DmaSupportTx: DmaSupport {
+        type TX: Tx;
+
+        fn with_tx<R, F: FnOnce(&mut Self::TX) -> R>(&mut self, f: F) -> R;
+    }
+
+    pub trait DmaSupportRx: DmaSupport {
+        type RX: Rx;
+
+        fn with_rx<R, F: FnOnce(&mut Self::RX) -> R>(&mut self, f: F) -> R;
+    }
+}
+
+/// DMA transaction for TX only transfers
+#[non_exhaustive]
+#[must_use]
+pub struct DmaTransferTx<'a, I>
+where
+    I: dma_private::DmaSupportTx,
+{
+    instance: &'a mut I,
```

Comment:
I have ideas how to avoid duplicating everything when introducing a Move-API but would need to explore that in more detail - feel free to do the same on top of this (I'm not yet too happy with my own ideas)